### PR TITLE
Major refactoring of the SeaClouds API / Proxy (Dashboard backend). IMPORTANT: This will break Angular.js frontend compatibility until it will updated to match the changes on SeaClouds API.

### LINKED ISSUE #52: Replace 3rd party services with mocks on dashboard tests
As said on https://github.com/SeaCloudsEU/SeaCloudsPlatform/pull/51, we should avoid using 3rd party non-reliable services for testing. 

Possible alternatives are mocks like: 
* https://github.com/square/okhttp/tree/master/mockwebserver
* http://www.mock-server.com/

### PR DESCRIPTION
**Pull request by adriannieto**
Major refactoring of the SeaClouds API / Proxy (Dashboard backend). IMPORTANT: _This will break Angular.js frontend compatibility until it will updated to match the changes on SeaClouds API_. It also includes changes from #181 in order to have a clean version of the backend until #180 is fixed. It fixes #52.

Due the size of the PR this review cannot be assumed by one person, so I will share the "review responsibility" across the involved partners. For this reason, the * API Proxy should be reviewed by the "Proxied" Component responsible to verify that the calls and the tests are matching the expected output*:
- @rosogon Please review *eu.seaclouds.platform.dashboard.proxy.SlaProxy*.
- @szenzaro Please review *eu.seaclouds.platform.dashboard.proxy.PlannerProxy*
- @MicheleGuerriero Please review *eu.seaclouds.platform.dashboard.proxy.MonitorProxy*
- @andreaturli  Please review *eu.seaclouds.platform.dashboard.proxy.DeployerProxy*.

In addition to this review, I would like to assign @andreaturli as the main reviewer.
 
Backend Improvements:
- Removed custom eu.seaclouds.platform.dashboard.util.http library with the Dropwizard Jersey standard one.
- Moved eu.seaclouds.platform.dashboard.config to eu.seaclouds.platform.dashboard.proxy
- Moved eu.seaclouds.platform.dashboard.resources package eu.seaclouds.platform.dashboard.rest
- Moved the proxy logic (HTTP Request redirection mainly) from eu.seaclouds.platform.dashboard.rest.* to eu.seaclouds.platform.dashboard.proxy.*
- Moved legacy monitoring (Brooklyn based one) from DeployerResource to MonitorResource
- Deleted inconsistent REST resources, for example if SLA are added from the Tosca DAM allowing a developer to call externally the SLA module to add unrelated SLA  Agreements could lead into inconsistencies.
- Imported T4C, SLA and Brooklyn Object Model.
- Added SeaCloudsApplicationDataStorage, a persistent layer based on Mapdb.
- Added SeaCloudsApplicationData model, which holds all the necesary information to retrive the related information and metadata (like Agreement or MonitoringRules) per each SeaClouds Application.
- Added ObjectMapperHelper.class to manually select the parser according to the Object class. This is neccesary due T4C, SLA and Brooklyn requires diferent parsers to work properly.

API Changes:
- Listing all the applications retrieves now a list of SeaCloudsApplicationData, This doesn't contain any information about the deployment status, if the user wants to retrieve information from a particular application it needs to call GET api/deployer/applications/seacloudsId. Please
- Now all the API methods use the SeaCloudsApplicationId instead of Brooklyn/SLA/T4C ID's.

Unit and Integration testing:
- Added basic Healthchecks to Dropwizard.
- Added unit tests to SeaCloudsApplicationData and SeaCloudsApplicationDataStorage.
- Added MockWebServer based tests to the API Proxy.
- Added Mockito based tests to the REST Resources.
- Added fixtures with the expected response of the Proxied services.Major refactoring of the SeaClouds API / Proxy (Dashboard backend). IMPORTANT: This will break Angular.js frontend compatibility until it will updated to match the changes on SeaClouds API.

### CODE REVIEW

**Comment by adriannieto** [comment]
There is an error looking for T4C dependency, maybe Travis cannot find the dependency?

```
[ERROR] Failed to execute goal on project dashboard: Could not resolve dependencies for project eu.seaclouds-project:dashboard:jar:0.8.0-SNAPSHOT: Failure to find it.polimi.tower4clouds:rules:jar:0.2.3 in http://repo.maven.apache.org/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]
```

**Comment by andreaturli** [comment]
great start @adriannieto 
I love your *ResourceTest!

I've not finished yet with the review, I've superficially looked at the overall structure: I think there is still room for simplification.
wdyt?

**Comment by adriannieto** [comment]
@andreaturli I was thinking about your idea of moving the clients to their respective module. The fact is that Modaclouds, Brooklyn and SLA ( not sure about  this @rosogon ) has their own Java client. However, the main issue I found was related with Maven dependency collision.

I think that importing their clients will require a lot of ```<exclusion></exclusion>``` magic.

**Comment by MicheleGuerriero** [comment]
@adriannieto I had a look at the MonitorProxy and it looks ok. Thanks!

**Comment by jacopogiallo** [comment]
@adriannieto I have checked the PlannerProxy and the corresponding PlannerProxyTest, and I see no issues in it. :+1:

**Comment by adriannieto** [comment]
Do not merge yet, I'm having some issues with the JSON parsing of the SLA related objects.

File: dashboard/src/main/java/eu/seaclouds/platform/dashboard/proxy/SlaProxy.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+     * @param agreementId of the SLA Agreement to be removed. This ID may differ from SeaClouds Application ID
+     * @return String representing that the Agreement was removed  properly
+     */
+    public String removeAgreement(String agreementId) {
+        Invocation invocation = getJerseyClient().target(getEndpoint() + "/agreements/" + agreementId).request()
+                .header("Accept", MediaType.APPLICATION_JSON)
+                .header("Content-Type", MediaType.APPLICATION_JSON)
+                .buildDelete();
+
+        //SLA Core returns a text message if the response was succesfully not the object, this is not the best behaviour
```
Comment (rosogon) [inline dashboard/src/main/java/eu/seaclouds/platform/dashboard/proxy/SlaProxy.java]: As specified in docs, the status code is the result. The response body is useless.

**Comment by adriannieto** [comment]
New bug detected due an outdated fixture:

```
java.lang.IllegalArgumentException: Not found MonitoringMetrics for 'throughput'
	at eu.seaclouds.platform.planner.aamwriter.MonitoringMetrics.from(MonitoringMetrics.java:34)
	at eu.seaclouds.platform.planner.aamwriter.Translator.buildModuleQoSRequirements(Translator.java:363)
	at eu.seaclouds.platform.planner.aamwriter.Translator.buildGroups(Translator.java:247)
	at eu.seaclouds.platform.planner.aamwriter.Translator.translate(Translator.java:58)
	at eu.seaclouds.platform.planner.aamwriter.AamWriter.writeAam(AamWriter.java:64)
	at eu.seaclouds.platform.planner.aamwriter.AamWriter.writeAam(AamWriter.java:58)
	at eu.seaclouds.platform.dashboard.rest.AamWriterResource.translateDesignerModel(AamWriterResource.java:48)
	at eu.seaclouds.platform.dashboard.rest.AamWriterResourceTest.testTranslateDesignerModel(AamWriterResourceTest.java:30)
```

Please @rosogon could you fix this?

File: dashboard/src/main/java/eu/seaclouds/platform/dashboard/proxy/SlaProxy.java (hunk 1 above)
Comment (adriannieto) [inline dashboard/src/main/java/eu/seaclouds/platform/dashboard/proxy/SlaProxy.java]: I will handle this after the fixtures will be updated.

**Comment by rosogon** [comment]
@adriannieto : Is this PR still in progress?

**Comment by adriannieto** [comment]
@rosogon  I think I can lock the status of this PR to finalize the review.

**Comment by rosogon** [comment]
@adriannieto : Squash, please.

**Comment by rosogon** [comment]
LGTM. Merging.

**Pull request merged**